waveorder/io/reader.py

```diff
@@ -0,0 +1,158 @@
+import numpy as np
+import os
+import zarr
+from tifffile import TiffFile
+
+
+class MicromanagerReader:
+
+    def __init__(self, folder):
+        """
+        loads ome-tiff files in folder into zarr or numpy arrays
+        Strategy:
+            1. read all files in folder
+            2. search for the file whose omexml does not contain the element tag "BinaryOnly"-> this is the master ome-tiff
+            3. each series represents a micro-manager stage position.  upon call for data, assign this data to a dict
+
+        :param folder: folder containing all ome-tiff files
+        :param reader: tifffile or aicsimageio
+        """
+
+        self._positions = {}
+        self.mm_meta = None
+        self.stage_positions = 0
+        self.height = 0
+        self.width = 0
+        self.frames = 0
+        self.slices = 0
+        self.channels = 0
+        self.chnames = []
+
+        self.master_ome_tiff = self._get_master_ome_tiff(folder)
+        print(self.master_ome_tiff)
+        self._set_mm_meta()
+
+    def _set_mm_meta(self):
+        with TiffFile(self.master_ome_tiff) as tif:
+            self.mm_meta = tif.micromanager_metadata
+
+            if self.mm_meta['Summary']['Positions'] > 1:
+                self.stage_positions = []
+                for p in range(self.mm_meta['Summary']['Positions']):
+                    print(f"appending {self.mm_meta['Summary']['StagePositions'][p]}")
+                    self.stage_positions.append(self.mm_meta['Summary']['StagePositions'][p])
+
+            for ch in self.mm_meta['Summary']['ChNames']:
+                print(f"appending {ch}")
+                self.chnames.append(ch)
+
+            self.height = self.mm_meta['Summary']['Height']
+            self.width = self.mm_meta['Summary']['Width']
+            self.frames = self.mm_meta['Summary']['Frames']
+            self.slices = self.mm_meta['Summary']['Slices']
+            self.channels = self.mm_meta['Summary']['Channels']
+
+    def _extract_data(self, master_ome):
+        """
+        extract all series from ome-tiff and place into dict of (pos: zarr)
+        :param master_ome: full path to master OME-tiff
+        :return:
+        """
+        with TiffFile(master_ome) as tif:
+            for idx, tiffpageseries in enumerate(tif.series):
+                self._positions[idx] = zarr.open(tiffpageseries.aszarr(), mode='r')
```

Comment:
This will create zarr stores for all series in the given ome.tif. This can create unnecessary overhead if you've only asked for one position. On the IBM clusters currently creating a zarr store with `store = tifffile.imread(path, aszarr=True)` takes an inordinate amount of time, tho I haven't tried the syntax you're using here.

Reply:
From what I understand, the major overhead occurs with calls to `tif.series`.  In Tifffile's code, it appears this call parses through the entire ome structure to get the page locations of each scene, then returns the list of pages (`tiffpageseries` above).

I could try assigning the ZarrTiffStore to self._positions[idx] rather than the zarr array.  But I found the memory footprint was low using the zarr array, and so assumed there was lazy loading anyways.

Reply:
Ok after some testing, I can't find any obvious way to interface with `tifffile` library to speed this up.  Any call to `tif.series` results in a calls to `_load` and then `_seek`, which scans all pages to identify their offset locations.  I'm not sure there's an easy solution here, as even aicsimageio uses tifffile to access data this way.

Reply:
Ah, I see, I didn't know that about `tif.series`. My suggestion here was to simply remove the for loop over tif.series and create a zarr store only for the series we're currently working with.

On that note, I'm not sure we should be treating the position dimension in the same way as all other dimensions. As you know, data in different series can have different dimensions, tho on the other hand MM acquisition doesn't really allow you to write data that way. Using `pycromanager` it is possible to write data of different dimensions at separate positions, but then again here we're developing a reader that's pretty tailored to MicroManager.

I think it'll be a good to separate the position dimensions from the rest. We can have methods like `set_series` to specify which position you want data for. We can also have `get_number_of_series` which will return the number of positions. That seems to be the approach that `aicsimageio` is taking in API v4.0 (https://github.com/AllenCellModeling/aicsimageio/tree/main). What do you think?

Reply:
>I think it'll be a good to separate the position dimensions from the rest. We can have methods like set_series to specify which position you want data for. 

`self._positions` is a dictionary whose keys are scene/position indicies.  So this would enable each key's value to have its own dimensions.

>We can also have get_number_of_series which will return the number of positions. That seems to be the approach that aicsimageio is taking in API v4.0 (https://github.com/AllenCellModeling/aicsimageio/tree/main). What do you think?

Ah!  I just removed this method but can put it back.  Li-Hao had a question about its default value when the data is not multiposition.

Reply:
I still don't think we should have a for loop over the series (`for idx, tiffpageseries in enumerate(tif.series)`). This create unnecessary overhead, when in practice we most often analyze one position at a time, and it's uncommon to switch back and forth between reading data from different positions, i.e. we sequentially go through all positions.

I think we should have a `set_position` or `set_series` function which creates a zarr store for the series/position the user wants to analyze. Call to this function should be optional, and if not called (the user calls `get_array` before `set_position`), it should default to series=0. We should also have a matching `get_series` so the user can confirm which series they're currently working with.

In this case one workflow will look like that:
```
data_obj = MicromanagerReader('data_path')
data = data_obj.get_array()
## analyze data
```
This will not create zarr stores for other positions in `data_path`.

Another workflow will look like this:
```
data_obj = MicromanagerReader('data_path')
data_obj.set_position(0)
data = data_obj.get_array() ## or data = data_obj.get_array(position=0)
## analyze data

data_obj.set_position(1)
data = data_obj.get_array()
## analyze data
```

`_create_zarr_store` will contain a section something like that:
```
def _create_zarr_store(self, position):
# self._position is the index of the current position/series
# self.zarr_store contains the zarr store
    if self._position != position:
        self._position = position
        self.zarr_store = zarr.open(tif_series, mode = 'r')
```

Reply:
I don't think it's possible to index directly into a position without first looping over the whole series -- this is the `tifffile` inefficiency.  I could be wrong and will look into this more.

But even if we must loop over the series, we can enable selective position calls still.  It'll just mean each position call can take variable amounts of time.

Reply:
Hm, interesting, I didn't know that. I thought you could create a TiffSeries object, but I see that that's not a thing. Ok,  I know you've looked a lot more into this, we'll go with your recommendation, you're welcome to resolve this issue

Reply:
I think your suggestion is a good one though.  It doesn't make sense to index every position when we just want one or two.  Am finishing some other repo dev and will get to this today :-)

Reply:
After thinking about this some, it might be very difficult to speed up calls to `tif.series` Likely we would need to clone and re-implement the `tifffile` library.  Let's leave this for later testing/research/